framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java

```diff
@@ -216,47 +184,124 @@ private static String replaceParameterSyntax(String expression) {
      * A visitor class that converts a JavaParser {@link Expression} to a {@link JavaExpression}.
      */
     private static class ExpressionToJavaExpressionVisitor
-            extends GenericVisitorWithDefaults<JavaExpression, JavaExpressionContext> {
+            extends GenericVisitorWithDefaults<JavaExpression, Void> {
 
         /**
          * The underlying javac API used to convert from Strings to Elements requires a tree path
          * even when the information could be deduced from elements alone. So use the path to the
          * current CompilationUnit.
          */
         private final TreePath pathToCompilationUnit;
+
         /** If non-null, the expression is parsed as if it were written at this location. */
         private final @Nullable TreePath localVarPath;
+
         /** The processing environment. */
         private final ProcessingEnvironment env;
+
         /** The resolver. Computed from the environment, but lazily initialized. */
         private @MonotonicNonNull Resolver resolver = null;
+
         /** The type utilities. */
         private final Types types;
 
         /** The java.lang.String type. */
-        TypeMirror stringTypeMirror;
+        private final TypeMirror stringTypeMirror;
+
+        /** The enclosing type. Used to look up unqualified method, field, and class names. */
+        private final TypeMirror enclosingType;
+
+        /**
+         * The expression to use for "this". If {@code null}, a parse error will be thrown if "this"
+         * appears in the expression.
+         */
+        private final @Nullable ThisReference thisReference;
+        /**
+         * For each formal parameter, the expression to which to parse it. For example, the second
+         * (index 1) element of the list is what "#2" parses to. If this field is {@code null}, a
+         * parse error will be thrown if "#2" appears in the expression.
+         */
+        private final @Nullable List<FormalParameter> parameters;
 
         /**
          * Create a new ExpressionToJavaExpressionVisitor.
          *
-         * @param pathToCompilationUnit required to use the underlying Javac API
+         * @param enclosingType type of the class that encloses the JavaExpression
+         * @param thisReference JavaExpression to which to parse "this", or null if "this" should
+         *     not appear in the expression
+         * @param parameters list of JavaExpressions to which to parse a formal parameter reference
+         *     such as "#2", or null if parameters should not appear in the expression
          * @param localVarPath if non-null, the expression is parsed as if it were written at this
          *     location
+         * @param pathToCompilationUnit required to use the underlying Javac API
          * @param env the processing environment
          */
         ExpressionToJavaExpressionVisitor(
```

Comment:
I suggest making this private.  It is only called by `convert` immediately below.